### PR TITLE
icon support in badge

### PR DESCRIPTION
**Pull request by adids1221**
## Description
`Icon` support in badge.
Added 2 new props `badgeProps | hideBadge` to the `Icon` component.

## Changelog
`Icon` support in badge.

### CODE REVIEW

**Comment by adids1221** [comment]
@Inbal-Tish this PR is mentioned in related PR in the private.
Take a note I changed the `size` type as part of `TODO` I found in the private repo.

File: src/components/icon/index.tsx [hunk 1]
```diff
@@ -18,6 +21,14 @@ export type IconProps = Omit<ImageProps, 'source'> &
...
      */
     assetGroup?: string;
+    /**
+     * Badge props passed down to Badge component
+     */
+    badgeProps?: BadgeProps;
+    /**
+     * Hide badge shown on button
+     */
+    hideBadge?: boolean;
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:31]: I don't see why this prop is needed... Render badge only if there is a badgeProps prop

File: demo/src/screens/componentScreens/IconScreen.tsx [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
-import {Assets, View, Icon, Text, Slider, Switch, GradientSlider} from 'react-native-ui-lib';
+import {Assets, View, Icon, Text, Slider, Switch, GradientSlider, Colors} from 'react-native-ui-lib';
 
 const IconScreen = () => {
   const [size, setSize] = useState(24);
   const [customSize, setCustomSize] = useState(false);
   const [color, setColor] = useState<string | number>();
   const [customColor, setCustomColor] = useState(false);
+  const [useBadge, setUseBadge] = useState(false);
+  const [usePimpleBadge, setUsePimpleBadge] = useState(false);
```
Comment (Inbal-Tish) [inline demo/src/screens/componentScreens/IconScreen.tsx:10]: The name of the component is 'pimple', not 'pimpleBadge'

File: demo/src/screens/componentScreens/IconScreen.tsx [hunk 3]
```diff
@@ -36,6 +38,20 @@ const IconScreen = () => {
...
       <GradientSlider type={GradientSlider.types.HUE} color={color as string} onValueChange={setColor}/>
       <Text marginT-s2>Custom color: {color || '#000000'}</Text>
+
+      <View marginV-s5>
+        <Text marginR-s2 text80BO>
+          Badge Settings:
+        </Text>
+        <View marginT-s3 row>
+          <Text marginR-s2>Use Badge:</Text>
+          <Switch value={useBadge} onValueChange={setUseBadge}/>
```
Comment (Inbal-Tish) [inline demo/src/screens/componentScreens/IconScreen.tsx:48]: renderBooleanOption function will render both Text and Switch from ExampleScreenPresenter

File: demo/src/screens/componentScreens/IconScreen.tsx [hunk 4]
```diff
@@ -36,6 +38,20 @@ const IconScreen = () => {
       </View>
       <GradientSlider type={GradientSlider.types.HUE} color={color as string} onValueChange={setColor}/>
       <Text marginT-s2>Custom color: {color || '#000000'}</Text>
+
+      <View marginV-s5>
+        <Text marginR-s2 text80BO>
+          Badge Settings:
```
Comment (Inbal-Tish) [inline demo/src/screens/componentScreens/IconScreen.tsx:44]: Don't you want to also display the options to change the icon and badge sizes?

File: src/components/icon/index.tsx [hunk 5]
```diff
@@ -83,7 +130,21 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
 
-  return isSvg(source) ? renderSvg() : renderImage();
+  return (
+    <View>
+      {isSvg(source) ? renderSvg() : renderImage()}
+      {shouldShowBadge && (
+        <Badge
+          pointerEvents={'none'}
+          {...badgeProps}
+          label={getBadgeLabel()}
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:140]: why not just pass the label from the badgeProps? Why the getter?

File: src/components/icon/index.tsx [hunk 6]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
+    } else {
+      badgePosition.right = -5;
+      badgePosition.top = -5;
+    }
+    return [badgePosition, containerStyle];
+  };
+
+  const getBadgeLabel = () => {
+    const {badgeProps} = props;
+    const label = get(badgeProps, 'label');
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:95]: why not optional chaining? badgeProps?.label

File: src/components/icon/index.tsx [hunk 7]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
+      badgePosition.top = -2;
+    } else {
+      badgePosition.right = -5;
+      badgePosition.top = -5;
+    }
+    return [badgePosition, containerStyle];
+  };
+
+  const getBadgeLabel = () => {
+    const {badgeProps} = props;
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:94]: why extract the badgeProps here again? This is a function component, we extract the props at the top

File: src/components/icon/index.tsx [hunk 8]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
   const {margins} = modifiers;
+  const shouldShowBadge = !hideBadge && badgeProps;
   const iconSize = size ? {width: size, height: size} : undefined;
   const shouldFlipRTL = supportRTL && Constants.isRTL;
 
+  const getBadgeStyling = (): StyleProp<ViewStyle> => {
+    const containerStyle = badgeProps?.containerStyle;
+    const badgePosition: StyleProp<ViewStyle> = {
+      position: 'absolute',
+      top: -3
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:81]: why did you set it to -3 if you're changing it later? I'd set the default here, for the badge, and if it is a pimple override.

File: src/components/icon/index.tsx [hunk 9]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
+    };
+    if (isPimple()) {
+      badgePosition.right = -2;
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:84]: these values should be a const. Are they not supposed to be dynamic or they should be the same for all icon/badge sizes?

File: src/components/icon/index.tsx [hunk 10]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
+  };
+
+  const getBadgeLabel = () => {
+    const {badgeProps} = props;
+    const label = get(badgeProps, 'label');
+    return label;
+  };
+
+  const getBadgeSize = () => {
+    const size = isPimple() ? 10 : 16;
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:100]: Why are you setting a default size? The badge and the pimple already have a default, and the user can pass it in the badgeProps if needed

File: src/components/icon/index.tsx [hunk 11]
```diff
@@ -54,12 +65,48 @@ const Icon = forwardRef((props: Props, ref: any) => {
...
+    return badgeProps?.size ? badgeProps?.size : size;
+  };
+
+  const isPimple = () => {
```
Comment (Inbal-Tish) [inline src/components/icon/index.tsx:104]: const isPimple = badgeProps && !badgeProps.label

File: src/components/icon/index.tsx (hunk 5 above)
Comment (adids1221) [reply src/components/icon/index.tsx:140]: you are right, I used the logic from the private repo - `BadgeButton` logic.

File: src/components/icon/index.tsx (hunk 9 above)
Comment (adids1221) [reply src/components/icon/index.tsx]: In the last commit I added support for dynamic badge size as we spoke.

**Review by Inbal-Tish — APPROVED** [review]

**Pull request merged**